modules/host/update.go

```diff
@@ -98,11 +107,127 @@ func (h *Host) threadedCreateStorageProof(obligation *contractObligation) {
 		h.log.Printf("ERROR: could not submit storage proof txn for %v (%v): %v", obligation.ID, obligation.Path, err)
 		return
 	}
+}
 
-	// Storage proof was successful, so increment profit tracking
-	h.mu.Lock()
-	h.profit = h.profit.Add(obligation.FileContract.Payout)
-	h.mu.Unlock()
+// handleActionItem looks at a contract obligation and contextually determines
+// if any actions need to be taken. Potential actions include submitting
+// storage proofs, resubmitting file contracts, and deleting the obligation.
+// handleActionItem will properly queue up any future actions that need to be
+// taken.
+func (h *Host) handleActionItem(co *contractObligation) {
+	// Check that the obligation has not already been deleted.
+	_, exists := h.obligationsByID[co.ID]
+	if !exists {
+		return
+	}
+
+	// Check for indications that the obligation should be deleted.
+	if !co.txnsConfirmed() && co.windowStart() < h.blockHeight {
+		// The final file contract revision has not made it into the
+		// blockchain, and the window for submitting it has closed. The
+		// obligation can be deleted.
+		h.removeObligation(co, obligationFailed)
+		return
+	}
+	if !co.proofConfirmed() && co.windowEnd() < h.blockHeight {
+		// The storage proof has not been confirmed and the window for
+		// submitting the proof has closed. The obligation can be deleted.
+		h.removeObligation(co, obligationFailed)
+		return
+	}
+	if co.proofConfirmed() && co.windowEnd()+confirmationRequirement <= h.blockHeight {
+		// The storage proof has been successful, and has enough confirmations
+		// to be considered stable. Therefore the obligation can be deleted.
+		h.removeObligation(co, obligationSucceeded)
+		return
+	}
```

Comment:
perhaps, though I'm not a huge fan of the idea of inserting so much logic and comments into a switch statement.